openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -412,30 +419,56 @@ def _verify_user(request, response, userid_in_session):
                 # If a view overrode the request.user with a masqueraded user, this will
                 #   revert/clean-up that change during response processing.
                 request.user = request.user.real_user
-            # The user at response time is expected to be None when the user
-            # is logging out.  We won't log that.
-            if request.safe_cookie_verified_user_id != request.user.id and request.user.id is not None:
-                log.warning(
-                    (
-                        "SafeCookieData user at request '{}' does not match user at response: '{}' "
-                        "for request path '{}'"
-                    ).format(  # pylint: disable=logging-format-interpolation
-                        request.safe_cookie_verified_user_id, request.user.id, request.path,
-                    ),
-                )
-                set_custom_attribute("safe_sessions.user_mismatch", "request-response-mismatch")
-            # The user session at response time is expected to be None when the user
-            # is logging out.  We won't log that.
-            if request.safe_cookie_verified_user_id != userid_in_session and userid_in_session is not None:
-                log.warning(
-                    (
-                        "SafeCookieData user at request '{}' does not match user in session: '{}' "
-                        "for request path '{}'"
-                    ).format(  # pylint: disable=logging-format-interpolation
-                        request.safe_cookie_verified_user_id, userid_in_session, request.path,
-                    ),
-                )
-                set_custom_attribute("safe_sessions.user_mismatch", "request-session-mismatch")
+
+            # determine if the request.user is different now than it was on the initial request
+            request_user_object_mismatch = request.safe_cookie_verified_user_id != request.user.id
+
+            # determine if the current session user is different than the user in the initial request
+            session_user_mismatch = request.safe_cookie_verified_user_id != userid_in_session
+
+            if request_user_object_mismatch or session_user_mismatch:
+                # Log accumulated information stored on request for each change of user
+                if hasattr(request, 'debug_user_changes'):
+                    log.info('\n'.join(request.debug_user_changes))
```

Comment:
We won't be able to reuse assert_no_warning logged because we will still log the top-level warning even if this more verbose logging is turned off, and that method fails if any log.warn occurs.